agent/eni/watcher/watcher_windows.go

```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "error occurred while instantiating watcher")
+	}
+
+	return &ENIWatcher{
+		ctx:              derivedContext,
+		cancel:           cancel,
+		agentState:       state,
+		eniChangeEvent:   stateChangeEvents,
+		primaryMAC:       primaryMAC,
+		interfaceMonitor: infMonitor,
+		notifications:    notificationChannel,
+		netutils:         networkutils.GetNetworkUtils(),
+	}, nil
+}
+
+// This method is used to reconcile the state of the ENIs attached to the instance
+// We get a map of all the interfaces and then reconcile the state of each interface
+func (eniWatcher *ENIWatcher) reconcileOnce() error {
+	state, err := eniWatcher.getAllInterfaces()
+	if err != nil {
+		return err
+	} else if state == nil || len(state) == 0 {
+		return nil
+	}
+
+	for mac := range state {
+		if err := eniWatcher.sendENIStateChange(mac); err != nil {
+			// This is not an error condition as the eni status message has already been sent
+			if strings.Contains(err.Error(), eniStatusSentMsg) {
+				continue
+			} else if _, ok := err.(*unmanagedENIError); ok {
+				log.Debugf("Windows ENI watcher reconciliation: unable to send state change: %v", err)
+			} else {
+				log.Warnf("Windows ENI watcher reconciliation: unable to send state change: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// This method is used to continuously monitor the notification channel for any notification
+// If any new notification is received then we will retrieve the MAC address of that interface and send the ENI state change event
+// If the context is cancelled then we will stop the monitor
+func (eniWatcher *ENIWatcher) eventHandler() {
+	for {
+		select {
+		case event := <-eniWatcher.notifications:
+			if event == iphelperwrapper.InitialNotification {
+				continue
+			}
+			go func(ctx context.Context, timeout time.Duration) {
+				mac, err := eniWatcher.netutils.GetInterfaceMACByIndex(event, ctx, timeout)
+				if err != nil {
+					log.Debugf("Windows ENI Watcher : event-handler : Cannot retrieve MAC Address for the interface with index : %v", event)
+					return
+				}
+
+				if err := eniWatcher.sendENIStateChangeWithRetries(ctx, mac, timeout); err != nil {
+					log.Warnf("Windows ENI Watcher : event-handler : Unable to send ENI State Change : %v", err)
+					return
+				}
+
+			}(eniWatcher.ctx, sendENIStateChangeRetryTimeout)
+		case <-eniWatcher.ctx.Done():
+			if err := eniWatcher.interfaceMonitor.Close(); err != nil {
+				log.Errorf("Windows ENI Watcher : event-handler : Cannot close the Windows interface monitor : %v", err)
+			}
+			return
+		}
+	}
+}
+
+// This method is used to obtain a list of all available valid interfaces
+func (eniWatcher *ENIWatcher) getAllInterfaces() (state map[string]int, err error) {
+	interfaces, err := eniWatcher.netutils.GetAllNetworkInterfaces()
+	if err != nil {
+		return nil, errors.Wrap(err, "error occurred while retrieving details of available interfaces.")
```

Comment:
errors.Wrap will render the message as `<second arg>:<first arg>.Err()` so don't have a period at the end of second arg

Reply:
My bad. I had missed this one.